### PR TITLE
[batch] Implement job run conditions

### PR DESCRIPTION
**Pull request by jigold**
CHANGELOG: Added `Job.run_when()` which allows you to specify under what conditions a job should run based on the state of its parent dependencies and whether the job is cancellable.

The gist of this PR is that we now have an extra variable `run_condition` which specifies under what conditions a job should run (any, all, always). The default is to maintain the current behavior that a job runs when all its parent dependencies succeeded. The semantics of `always_run` should stay the same.

Can you check over the jobs_after_update trigger and make sure I'm not crazy and nothing needs to be modified there since the job state where the cancelled state matters is a Ready or Running job and the run_condition / cancelled interaction applies to Pending jobs with parent dependencies.

Also, I did this crazily fast, so maybe I'm missing something and it shouldn't be this easy????

### CODE REVIEW

File: batch/sql/estimated-current.sql [hunk 1]
```diff
@@ -1049,7 +1050,10 @@ BEGIN
                jobs.job_id = t.job_id
           SET jobs.state = IF(COALESCE(t.n_pending_parents, 0) = 0, 'Ready', 'Pending'),
               jobs.n_pending_parents = COALESCE(t.n_pending_parents, 0),
-              jobs.cancelled = IF(COALESCE(t.n_succeeded, 0) = COALESCE(t.n_parents - t.n_pending_parents, 0), jobs.cancelled, 1)
+              jobs.cancelled = IF(jobs.run_condition = 'all',
+                                  IF(COALESCE(t.n_succeeded, 0) = COALESCE(t.n_parents - t.n_pending_parents, 0), jobs.cancelled, 1),
+                                  IF(jobs.run_condition = 'always', 0,
+                                     (COALESCE(t.n_parents, 0) = 0) | COALESCE(t.n_succeeded, 0) > 0)  # run_condition = 'any'
```
Comment (daniel-goldstein) [inline batch/sql/estimated-current.sql:1056]: Nit: Just for ease of readability, can we use a [CASE expression](https://www.w3schools.com/sql/func_mysql_case.asp) instead of the nested `IF`? Feels like it should map better to the all / any / always enum.

File: batch/batch/front_end/front_end.py [hunk 2]
```diff
@@ -1018,6 +1020,7 @@ async def _create_jobs(userdata: dict, job_specs: dict, batch_id: int, update_id
                 inst_coll_name,
                 n_regions,
                 regions_bits_rep,
+                run_condition == 'any' and len(parent_ids) > 0,
```
Comment (daniel-goldstein) [inline batch/batch/front_end/front_end.py:1023]: This took me a couple reads to get what's going on here and this is a clever way to get at the state change of `any` jobs. However, it feels pretty funky to me that `any` jobs (mostly) start out as cancelled and are switch to not-cancelled when a parent succeeds. It feels like there should be a distinction between the states

- "none of my parents have completed yet"
- "all my parents so far have failed"
- "I will never run because all my parents failed"

I also don't love that `n_cancelled` can potentially go down and that a job could get uncancelled (what if I hit the `cancel` button which sets the job as cancelled but then a succeeding parent uncancels it?

How much work would it be to do the following?

All jobs start as not cancelled. We track `n_succeeded_parents`, and then we can set an any job as cancelled if `n_parents > 0 & n_pending_parents = 0 & n_succeeded_parents = 0`?

**Review by daniel-goldstein — CHANGES_REQUESTED** [review]
Looks great. I have one concern about how `any` jobs are handled, but otherwise this looks totally reasonable. Amazing how small a change it is.

**Comment by jigold** [comment]
Do you think we should add 1-2 more states which are 'any_failure', 'all_failed' and change the others to 'any_success' and 'all_succeeded'? I was thinking it might be nice to have onError jobs with that dependency explicitly in the DAG.

**Comment by daniel-goldstein** [comment]
> Do you think we should add 1-2 more states which are 'any_failure', 'all_failed' and change the others to 'any_success' and 'all_succeeded'? I was thinking it might be nice to have onError jobs with that dependency explicitly in the DAG.

I'm fine with this, though I think `any_succeeded` and `all_succeeded` is more consistent (instead of `any_success`)

**Comment by daniel-goldstein** [comment]
Maybe we make the name change for success and put the failure conditions in a separate PR? I wonder if there are any sneaky bits with those additions

**Comment by jigold** [comment]
The test failures were unrelated to the proposed changes.

File: batch/sql/estimated-current.sql [hunk 3]
```diff
@@ -1531,7 +1541,15 @@ BEGIN
...
-          jobs.cancelled = IF(new_state = 'Success', jobs.cancelled, 1)
+          jobs.n_succeeded_parents = jobs.n_succeeded_parents + (new_state = 'Success'),
+          jobs.cancelled = (CASE jobs.run_condition
+                WHEN 'always' THEN jobs.cancelled
+                WHEN 'all_succeeded' THEN IF(new_state = 'Success', jobs.cancelled, 1)
+                WHEN 'any_succeeded' THEN IF(COALESCE(jobs.n_pending_parents - 1, 0) = 0,
+                                             IF(COALESCE(jobs.n_succeeded_parents, 0) + (new_state = 'Success') > 0, jobs.cancelled, 1),
+                                             jobs.cancelled)
+                ELSE IF(new_state = 'Success', jobs.cancelled, 1)  # backwards compatibility
+                END)
```
Comment (daniel-goldstein) [inline batch/sql/estimated-current.sql:1552]: Sorry I'm still finding this quite hard to parse. Is the following equivalent? I find it easier to frame the answer to the question as "The job is cancelled if \<some condition is met\> OR the job was already cancelled". This way we never uncancel a job which I feel like is an important property.

```
CASE jobs.run_condition
WHEN 'always' THEN jobs.cancelled
WHEN 'all_succeeded' THEN new_state != 'Success' OR jobs.cancelled
WHEN 'any_succeeded' THEN (jobs.n_pending_parents - 1 = 0 AND jobs.n_succeeded_parents = 0 AND new_state != 'Success') OR jobs.cancelled
ELSE new_state != 'Success' or jobs.cancelled
END
```

**Review by daniel-goldstein — CHANGES_REQUESTED** [review]

File: batch/sql/estimated-current.sql [hunk 4]
```diff
@@ -244,12 +244,14 @@ CREATE TABLE IF NOT EXISTS `jobs` (
   `cores_mcpu` INT NOT NULL,
   `status` TEXT,
   `n_pending_parents` INT NOT NULL,
+  `n_succeeded_parents` INT DEFAULT NULL,
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/sql/estimated-current.sql:247]: I was trying to make it so old records didn't have 0's for the n_succeeded_parents as that means something different than NULL to me and I thought it would be more error prone.

[the remark this answers is not in the document]
Comment (daniel-goldstein) [reply batch/sql/estimated-current.sql:247]: ah that's a good point, but does feel a bit hacky to leave like this forever. I feel like this deserves its own migration of INT DEFAULT NULL -> set to proper value -> INT NOT NULL. Can we do this?

[the remark this answers is not in the document]
Comment (jigold) [reply batch/sql/estimated-current.sql:247]: We theoretically can, but I was worried about how to make this efficient and plan the migration so it's fast, idempotent, and correct. I guess my original feeling was the work wasn't worth it for a field that's only used when a job hasn't completed yet, but if you feel differently, I can try and work out what exactly needs to happen to do the migration.

[the remark this answers is not in the document]
Comment (jigold) [reply batch/sql/estimated-current.sql:247]: @daniel-goldstein Can I get input on this? Do you want to add a migration?

[the remark this answers is not in the document]
Comment (daniel-goldstein) [reply batch/sql/estimated-current.sql:247]: Ya I think ultimately this will be a useful field, not just for this particular use case, so would prefer to introduce it the right way.

**Comment by jigold** [comment]
I'd like to test this on the real database just to make sure it all works first.

**Comment by jigold** [comment]
Closing for now until there is need for this feature or we want to pick it back up.
